[FSDP] Fix load_sharded_state_dict FQN mismatches for shared parameters

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86524

`_sharded_pre_load_state_dict_hook()` should calls `_param_fqns()` to ensure shared parameters names are also included.

Differential Revision: [D40201304](https://our.internmc.facebook.com/intern/diff/D40201304/)

fegin committed:
[FSDP] Fix load_sharded_state_dict FQN mismatches for shared parameters

`_sharded_pre_load_state_dict_hook()` should calls `_param_fqns()` to ensure shared parameters names are also included.

Differential Revision: [D40201304](https://our.internmc.facebook.com/intern/diff/D40201304/)

[ghstack-poisoned]

awgu reviewed (commented):
This change makes sense to me, but if we want to land this formally, should we include a unit test?

Her are some code examples for how to use `TransformerWithSharedParams.init()`, which has shared parameters (see `common_fsdp.py` for the precise details).

No FSDP:
https://github.com/pytorch/pytorch/blob/be682befbc836a07d5d070bb569450429526a64b/test/distributed/fsdp/test_fsdp_use_orig_params.py#L84-L89

FSDP:
https://github.com/pytorch/pytorch/blob/be682befbc836a07d5d070bb569450429526a64b/test/distributed/fsdp/test_fsdp_use_orig_params.py#L115-L133

fegin commented:
@awgu Yup, that makes sense. What is puzzling me is that I thought `test_state_dict` already has shared parameters test cases but somehow this issue is not caught. I'm looking into the UT to understand what is missing.

fegin commented:
@awgu  Scratch my comment. I didn't find shared parameters testing in `test_fsdp_state_dict`, probably was confused with Pyper's tests.

fegin committed:
Update on "[FSDP] Fix load_sharded_state_dict FQN mismatches for shared parameters"


`_sharded_pre_load_state_dict_hook()` should calls `_param_fqns()` to ensure shared parameters names are also included.

Differential Revision: [D40201304](https://our.internmc.facebook.com/intern/diff/D40201304/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -233,6 +233,36 @@ def test_fsdp_state_dict_with_activation_checkpoint(self, state_dict_type, check
+        )
+
+        fsdp_model = model_creator()
+        for tensor in itertools.chain(fsdp_model.parameters(), fsdp_model.buffers()):
```
what does this portion add to the unittest?

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2179,6 +2179,15 @@ def _param_fqns(self) -> Iterator[Tuple[str, str, str]]:
+    @property
+    def _shared_param_fqns(self) -> Iterator[Tuple[str, str, str]]:
+        for param_name, module_name in (
+            self._fsdp_wrapped_module.handle.shared_parameter_module_names()
```
nit: i remembered _fsdp_wrapped_module could have multiple handles, so this should be "self._fsdp_wrapped_module.handles[0]"? @awgu

fegin committed:
Update on "[FSDP] Fix load_sharded_state_dict FQN mismatches for shared parameters"


`_sharded_pre_load_state_dict_hook()` should calls `_param_fqns()` to ensure shared parameters names are also included.

Differential Revision: [D40201304](https://our.internmc.facebook.com/intern/diff/D40201304/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Fix load_sharded_state_dict FQN mismatches for shared parameters"


`_sharded_pre_load_state_dict_hook()` should calls `_param_fqns()` to ensure shared parameters names are also included.

Differential Revision: [D40201304](https://our.internmc.facebook.com/intern/diff/D40201304/)

[ghstack-poisoned]
